packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -987,14 +1111,30 @@ class SliverMultiBoxAdaptorElement extends RenderObjectElement implements Render
           _childElements.remove(index);
         }
       }
-      // processElement may modify the Map - need to do a .toList() here.
-      _childElements.keys.toList().forEach(processElement);
+
+      for (int index in _childElements.keys.toList()) {
+        final int newIndex = getNewIndex(_childElements[index].widget);
```

Comment:
I would probably move the content of getNewIndex into this function to make this slightly easier to read, e.g.

final Key key = _childElements[index].widget.key;
final int newIndex = key == null ? -1 : widget.delegate.findIndexByKey(w.key);